File: C++/93. Restore IP Addresses.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> res;
        helper(res, "", s, 0, 0);
        return res;
    }
    void helper(vector<string> &res, string p, string &s, int i,  int n) {
        if (i == s.size() && n==4) {
            p.pop_back();
            res.push_back(p);
            return;
        }
        if(n > 4){
            return;   
        }
        else{
            for(int j=i; j<i + 3 &&  j<(int)s.size(); j++) {
                if(stoi(s.substr(i, j - i + 1)) <= 255 && (i == j || s[i] != '0')) {
                    helper(res, p + s.substr(i, j - i + 1) + ".", s, j+1, n+1);
                }
            }
        }
    }
};

File: Java/93. Restore IP Addresses.java
class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> res = new ArrayList<>();
        helper(res, new ArrayList<>(), s, 0);
        return res;
    }
    public void helper(List<String> res, List<String> l, String s, int i) {
        int n = l.size();
        if (i >= s.length()) {
            if (n == 4) 
                res.add(String.join(".", l));
                
            return;
        }
        if(n >= 4){
            return;   
        }
        else{
            for (int j = 1; j <= 3 && i + j <= s.length(); j++) {
                String str = s.substring(i, i + j);
                if (Integer.parseInt(str) >= 0 && Integer.parseInt(str) <= 255) {
                    l.add(str);
                    helper(res, l, s, i + j);
                    l.remove(l.size() - 1);
                }
                if (Integer.parseInt(str) == 0){
                    break;
                }
            }
        }
    }
}
